test/test_functional_tensor.py

```diff
@@ -36,6 +36,29 @@ def test_crop(self):
         self.assertTrue(torch.equal(img_cropped, (img_cropped_GT * 255).to(torch.uint8)),
                         "functional_tensor crop not working")
 
+    def test_adjustments(self):
+        fns = ((F.adjust_brightness, F_t.adjust_brightness),
+               (F.adjust_contrast, F_t.adjust_contrast),
+               (F.adjust_saturation, F_t.adjust_saturation))
+
+        for _ in range(20):
+            channels = 3
+            dims = torch.randint(1, 50, (2,))
+            img = torch.rand(channels, *dims, dtype=torch.float)
+            factor = 3 * torch.rand(1)
+            for f, ft in fns:
+
+                ft_img = ft(img, factor)
+
+                img_pil = transforms.ToPILImage()(img)
+                f_img_pil = f(img_pil, factor)
+                f_img = transforms.ToTensor()(f_img_pil)
+
+                # F uses uint8 and F_t uses float, so there is a small
+                # difference in values caused by truncations.
+                l1_diff = (ft_img - f_img).norm(p=1)
+                self.assertLess(l1_diff, 0.01 * img.nelement())
```

Comment:
This sounds reasonable. From running the tests thousands of times, the highest error I got was 4 / 255; it seems theoretically possible to have a truncation error up 5 / 255 (it seems we have 5 multiplications followed by truncations), so I set that as a bound.

Reply:
So, I made the returned tensor to have the same dtype, but as it is we still have truncations, since internal additions still happen with floats. The simplest way to resolve this seems to be to do a tensor multiplication per element, convert the dtype, then sum-reduce the tensor.
That makes the code a bit less clear though, and since keeping the truncation doesn't seem to be very important, I kept as it is.
If you think it is worth to have truncation for uint8, I can still add it :)

Reply:
This is fine, but I'd love to understand the truncations you are mentioning. From the [Pillow implementation of Blend](https://github.com/python-pillow/Pillow/blob/710a5207bd961b165ba7f8dbc523f5bd8d93fe2c/src/libImaging/Blend.c#L57-L58), there is some casting to float happening, but maybe this is not the truncation that you mention.

Reply:
Yeah, so I was looking at the worst case in the 3 functions, which is in adjust_contrast, that you have 3 truncations in the grayscale conversion ( https://github.com/python-pillow/Pillow/blob/master/src/libImaging/Convert.c#L45 ), and then the truncations from the interpolation. I just assumed that their interpolation implementation had 2 truncations, but looking closely, it has only 1 - that explains why I only saw errors up to 4 / 255 in the float implementation, and up to 3 / 255 in the uint8 implementation :)

Reply:
My impression was that the grayscale conversion was doing 3 truncations and I assumed the interpolation was doing 2 - looking at the code, the interpolation only does 1 - which would explain why I only saw errors up to 4 / 255 in float (and up to 3 / 255 in uint8) :)